src/index/blockfilterindex.cpp

```diff
@@ -146,18 +146,19 @@ bool BlockFilterIndex::CustomCommit(CDBBatch& batch)
     return true;
 }
 
-bool BlockFilterIndex::ReadFilterFromDisk(const FlatFilePos& pos, const uint256& hash, BlockFilter& filter) const
+bool BlockFilterIndex::ReadFilterFromDisk(const FlatFilePos& pos, const uint256& hash, const uint256& block_hash, BlockFilter& filter) const
 {
     AutoFile filein{m_filter_fileseq->Open(pos, true)};
     if (filein.IsNull()) {
         return false;
     }
 
     // Check that the hash of the encoded_filter matches the one stored in the db.
-    uint256 block_hash;
+    uint256 read_block_hash;
     std::vector<uint8_t> encoded_filter;
     try {
-        filein >> block_hash >> encoded_filter;
+        filein >> read_block_hash >> encoded_filter;
+        if (read_block_hash != block_hash) return error("Checksum mismatch in filter decode.");
```

Comment:
This error would be easier to debug if it logs an specific error like "Block hash mismatch in filter decode". 

Plus, could leave this as it was before if the new arg is called `expected_block_hash` (no need to do the `read_block_hash` rename).